pdns/dnsdist-cache.cc

```diff
@@ -294,75 +294,80 @@ bool DNSDistPacketCache::get(DNSQuestion& dq, uint16_t queryId, uint32_t* keyOut
 
 /* Remove expired entries, until the cache has at most
    upTo entries in it.
+   If the cache has more than one shard, we will try hard
+   to make sure that every shard has free space remaining.
 */
-size_t DNSDistPacketCache::purgeExpired(size_t upTo)
+size_t DNSDistPacketCache::purgeExpired(size_t upTo, const time_t now)
 {
-  size_t removed = 0;
-  uint64_t size = getSize();
-
-  if (size == 0 || upTo >= size) {
-    return removed;
-  }
-
-  size_t toRemove = size - upTo;
+  const size_t maxPerShard = upTo / d_shardCount;
 
+  size_t removed = 0;
   size_t scannedMaps = 0;
 
-  const time_t now = time(nullptr);
   do {
     uint32_t shardIndex = (d_expungeIndex++ % d_shardCount);
```

Comment:
Since  `d_expungeIndex` is incremented `d_shardCount` times I believe the scanning will always start at shard 0. If you wamnt to start at a different shard for each `purge` call the increment needs to be above the loop. Or just removed `expungeIndex` ?

Reply:
Right, since we now always scan all buckets ` d_expungeIndex` is not relevant anymore, I'll remove it.